### PR TITLE
Update -bin header restrictions, and re introduce string interning.

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
/cc @ejona86

### CODE REVIEW

File: core/src/test/java/io/grpc/MetadataTest.java [hunk 1]
```diff
@@ -189,6 +193,13 @@ public void roundTripIntegerMarshaller() {
     roundTripInteger(0x87654321);
   }
 
+  @Test
+  public void invalidBinaryKeyName() {
```
Comment (nmittler) [inline core/src/test/java/io/grpc/MetadataTest.java]: You're only testing one of the conditions here ... should probably also test something that doesn't end with "-bin".

File: core/src/main/java/io/grpc/Metadata.java [hunk 2]
```diff
@@ -449,7 +449,8 @@ private static String validateName(String n) {
 
     private Key(String name) {
       this.originalName = checkNotNull(name);
-      this.name = validateName(this.originalName.toLowerCase(Locale.ROOT));
+      // Intern the result for faster string identity checking.
+      this.name = validateName(this.originalName.toLowerCase(Locale.ROOT)).intern();
```
Comment (buchgr) [inline core/src/main/java/io/grpc/Metadata.java]: Don't you also have to update `Key.equals` for that to work? Also wouldn't it be sufficient to do the interning in `Key.equals` when comparing?

Comment (buchgr) [inline core/src/main/java/io/grpc/Metadata.java]: or is this just to safe memory?

File: core/src/test/java/io/grpc/MetadataTest.java (hunk 1 above)
Comment (carl-mastrangelo) [inline core/src/test/java/io/grpc/MetadataTest.java]: Added another.

File: core/src/main/java/io/grpc/Metadata.java (hunk 2 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/Metadata.java]: No, Key.equals calls name.equals.  This is the String class equals which checks pointers before characters:

```
    public boolean equals(Object anObject) {
        if (this == anObject) {
            return true;
        }
        ...
```

**Comment by ejona86** [comment]
@carl-mastrangelo LGTM

**Pull request merged**